### PR TITLE
Fix big buffers (4GB+) overflow in ttnn codebase and metal slow dispatch

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/22336)

### Problem description
ttnn.from_torch / to_torch doesn't really work well for huge tensors with sizes not fitting in uint32_t. This PR is first step to fix that

### What's changed
Fix integer overflow in TTNN code base:
- tilizing
- padding

Fix integer overflow in metal code base for slow dispatch
- Most changed in `tt_metal/api/tt-metalium/buffer.hpp`

Added c++ test for tilizing of 4Gb+ buffer:
- `tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp`
- It's turned off by default, since it takes a few minutes to run. If you need to run, pass `--gtest_also_run_disabled_tests` to gtest binary

Added python test for moving to device and then back > 4Gb of data:
- `tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py`
- Add `slow` mark for ttnn pytests to mark slow tests, that won't run by default, unless `--runslow` is passed (see `tests/ttnn/conftest.py`)
- This tests takes a ton of time, since slow dispatch readback is extremely slow

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes -- https://github.com/tenstorrent/tt-metal/actions/runs/15819570114/job/44589848819
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) -- https://github.com/tenstorrent/tt-metal/actions/runs/15821968999
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/api/tt-metalium/buffer.hpp [hunk 1]
```diff
@@ -234,11 +234,13 @@ class Buffer final {
 
     // Returns address of buffer in the first bank
     uint32_t address() const;
+    uint64_t address_u64() const;
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/api/tt-metalium/buffer.hpp:237]: Can we use `DeviceAddr` type alias instead of uint64_t where u64 was added for consistency?
Also file an issue to clean up and replace the u32 methods entirely?
This is also related to https://github.com/tenstorrent/tt-metal/issues/14294 and the assertion in `Buffer::allocate_impl()` we should remove once this is done.

**Comment by bbradelTT** [comment]
Would all ops need to be incrementally updated to handle inputs over 4GB?

File: tt_metal/api/tt-metalium/buffer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (omilyutin-tt) [reply tt_metal/api/tt-metalium/buffer.hpp:237]: Let's make a `StrongType`  `DeviceAddr` that supports uint64. Spell `DeviceAddress` and migrate `DeviceAddr` to the new one, maybe?

File: tt_metal/api/tt-metalium/buffer.hpp [hunk 2]
```diff
@@ -256,9 +258,9 @@ class Buffer final {
 
     bool bottom_up() const { return bottom_up_; }
 
-    DeviceAddr page_address(uint32_t bank_id, uint32_t page_index) const;
+    DeviceAddr page_address(uint64_t bank_id, uint64_t page_index) const;
 
-    DeviceAddr bank_local_page_address(uint32_t bank_id, uint32_t page_index) const;
+    DeviceAddr bank_local_page_address(uint64_t bank_id, uint64_t page_index) const;
```
Comment (omilyutin-tt) [inline tt_metal/api/tt-metalium/buffer.hpp:263]: Why do these need to be migrated to u64? Just curious

Comment (philei-tt) [reply tt_metal/api/tt-metalium/buffer.hpp:263]: Just to reduce number of static_cast<uint64_t> in the code

File: ttnn/core/tensor/tensor_impl.cpp [hunk 3]
```diff
@@ -1254,10 +1254,24 @@ Tensor pad(
...
             std::memcpy(
-                output_buffer.data() + input_tensor_start[0], input_buffer.begin(), input_padded_shape[0] * sizeof(T));
+                output_buffer.data() + input_tensor_start[0],
+                input_buffer.begin(),
+                static_cast<size_t>(input_padded_shape[0]) * sizeof(T));
             return output_buffer;
         }
 
+        // Note: compute_strides from shape.hpp can overflow
+        auto compute_strides = [](const tt::tt_metal::Shape& shape) -> tt::stl::SmallVector<size_t> {
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply ttnn/core/tensor/tensor_impl.cpp:1264]: Yeah, makes sense, will do

File: tt_metal/api/tt-metalium/buffer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (philei-tt) [reply tt_metal/api/tt-metalium/buffer.hpp:237]: I will change to DeviceAddr and file issue for refactoring. I agree that making it StrongType would be nice, but that would require a gazillion changes, since there a many places where it's compared to literals and other types. This should probably be done in a scope of this future refactoring issue

**Comment by philei-tt** [comment]
@tt-aho @omilyutin-tt Issue for typing refactoring: https://github.com/tenstorrent/tt-metal/issues/22752

**Review by bbradelTT — APPROVED** [review]

**Comment by ThisIsFineTM** [comment]
Any thoughts on using `size_t` and mirroring how the standard library does it?

Are there any specific hardware limitations requiring the use of a 32 bit integer for anything?

**Comment by philei-tt** [comment]
> Any thoughts on using `size_t` and mirroring how the standard library does it?
> 
> Are there any specific hardware limitations requiring the use of a 32 bit integer for anything?

- You mean using size_t instead of uint64_t? 
- Not sure about hardware limitations, but right now all runtime and compile_time arguments have 32bits in size, so that's one of the limitations. Also, as I understood from @pgkeller fast dispatch has to work with uint32_t types sizes and offsets.

**Comment by mpiseTT** [comment]
I came upon this PR through https://github.com/tenstorrent/tt-metal/issues/22336. The prefetcher (device) code is not limited to 32-bit addressing because the memory is spread out over different NOC nodes. Thus a > 4 GB address is resolved by NOC address and address within the node (which does not exceed 32-bits). So this seems to be a Metal API issue alone, and might require the ability to handle 64-bit buffer size and offset values (i.e. not even address values) until where the CQ interface resolves to NOC address and NOC-relative address.

File: tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp [hunk 4]
```diff
@@ -589,3 +585,71 @@ INSTANTIATE_TEST_SUITE_P(
...
         ::testing::Bool()                                               // transpose
         ));
+
+// Note: tuple is used for ::testing::Combine
+using TilizeUntilizeBigBuffersParams = std::tuple<PhysicalSize, TensorLayoutType, TensorLayoutType, bool, bool>;
+
+class TilizeUntilizeBigBufferTestsFixture : public ::testing::TestWithParam<TilizeUntilizeBigBuffersParams> {};
+
+// Disabled by default, since they take a lot of timne to run. To still run them, use --gtest_also_run_disabled_tests
+TEST_P(TilizeUntilizeBigBufferTestsFixture, DISABLED_TilizeUntilizeBigBuffer) {
```
Comment (mpiseTT) [inline tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp]: Can you define default parameters for this test and not disable it? If it is left disabled then it might regress over time.

File: tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import ttnn
+import torch
+import pytest
+import os
+
+
+@pytest.mark.parametrize(
+    "shape",
+    [
+        (2**6, 2**13, 2**13 + 32),  # (4Gb + 32) * 2 (sizeof(bfloat16))
```
Comment (mpiseTT) [inline tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py]: Update the calculated sizes in comments? 64 x 8192 x (8192 + 32) =4,311,744,512 \ne 64 x 8192 x 8192 + 32 =4,294,967,328

File: tt_metal/api/tt-metalium/buffer.hpp [hunk 6]
```diff
@@ -226,11 +226,13 @@ class Buffer final : public std::enable_shared_from_this<Buffer> {
 
     // Returns address of buffer in the first bank
     uint32_t address() const;
+    DeviceAddr address_u64() const;
```
Comment (mpiseTT) [inline tt_metal/api/tt-metalium/buffer.hpp]: I think `Buffer::address()` is an on device address, relative to a specific core. If so, then these currently are 32-bit values. Please check, as I am relatively inexperienced in this code and hence could be wrong. Do your tests reveal this to be the limitation?

File: tt_metal/api/tt-metalium/tt_metal.hpp [hunk 7]
```diff
@@ -332,7 +332,7 @@ uint32_t EncodePerDeviceProgramID(uint32_t base_program_id, uint32_t device_id,
  * | [DRAM_UNRESERVED_BASE, dram_size)         | Yes      | | host_buffer  | Buffer on host to copy data from |
  * std::vector<uint32_t> | Host buffer must be fully fit DRAM buffer | Yes      |
  */
-bool WriteToDeviceDRAMChannel(IDevice* device, int dram_channel, uint32_t address, std::vector<uint32_t>& host_buffer);
+bool WriteToDeviceDRAMChannel(IDevice* device, int dram_channel, uint64_t address, std::vector<uint32_t>& host_buffer);
```
Comment (mpiseTT) [inline tt_metal/api/tt-metalium/tt_metal.hpp]: Same as above - was uint32_t address tied to any actual addressing limitation, since the `address` here is on core address.

File: tt_metal/api/tt-metalium/tt_metal.hpp [hunk 8]
```diff
@@ -350,7 +350,7 @@ bool WriteToDeviceDRAMChannel(IDevice* device, int dram_channel, uint32_t addres
  * | Buffer on host to copy data into                             | std::vector<uint32_t> | | Yes      |
  */
 bool ReadFromDeviceDRAMChannel(
-    IDevice* device, int dram_channel, uint32_t address, uint32_t size, std::vector<uint32_t>& host_buffer);
+    IDevice* device, int dram_channel, uint64_t address, uint32_t size, std::vector<uint32_t>& host_buffer);
```
Comment (mpiseTT) [inline tt_metal/api/tt-metalium/tt_metal.hpp]: Same

File: tt_metal/api/tt-metalium/buffer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpiseTT) [reply tt_metal/api/tt-metalium/buffer.hpp]: My overall comment is do all the addresses/sizes/offsets need to be 64-bit, or is some calculation somewhere overflowing, which must be identified and fixed? Are device buffer addresses required to be 64-bit?

File: tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    shape = (2**6, 2**13, 2**13 + 32)
+    if slow_dispatch is None or slow_dispatch == "0":
+        pytest.skip("Requires slow dispatch, skipping test")
+
+    torch_input = torch.full(shape, 1).bfloat16()
+    torch_output = torch_input
+
+    input_tensor = ttnn.from_torch(torch_input, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
+    output_tensor = ttnn.from_device(input_tensor)
+    output_tensor = ttnn.to_torch(output_tensor).bfloat16()
```
Comment (ayerofieiev-tt) [inline tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py:30]: are you aware about `ttnn.full`?

File: tt_metal/api/tt-metalium/buffer.hpp (hunk 6 above)
Comment (ayerofieiev-tt) [inline tt_metal/api/tt-metalium/buffer.hpp]: This is a weird API, not good. 
Can you explain whats going on here?

File: tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py (hunk 9 above)
Comment (philei-tt) [reply tests/ttnn/stress_tests/test_create_buffer_bigger_than_4gb.py:30]: This test is testing overflow happening in ttnn.from_torch, which also tests overflow in padding, tilizing and dispatching. AFAIK ttnn.full created buffer directly on device, which skips most of those steps

File: tt_metal/api/tt-metalium/buffer.hpp (hunk 6 above)
Comment (philei-tt) [reply tt_metal/api/tt-metalium/buffer.hpp]: So changing address return type to DeviceAddr is not an option since address is often passed to kernel arguments that rely on type deduction for std::array, so that would require static_cast in a ton of places. But as @mpiseTT noted, buffer represents data on device (BufferType::SYSTEM_MEMORY is not used anymore), where 32-bit indexing is enough, so I should probably nuke that and just add static_cast<BufferType> in a few places that overflow

Comment (philei-tt) [reply tt_metal/api/tt-metalium/buffer.hpp]: Yeah, all the kernels take address as uint32_t, and that won't be changed any time soon... You're right, better just cast place with overflow and remove that

File: tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp (hunk 4 above)
Comment (philei-tt) [reply tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp]: Added another test to test only row-major -> n-faces case. Takes ~27s on my machine

File: tt_metal/api/tt-metalium/buffer.hpp (hunk 6 above)
Comment (philei-tt) [reply tt_metal/api/tt-metalium/buffer.hpp]: Removed

File: tt_metal/api/tt-metalium/shape_base.hpp [hunk 10]
```diff
@@ -19,6 +19,14 @@ class ShapeBase {
...
     explicit ShapeBase(const Container& shape) : value_(shape) { init(); }
     explicit ShapeBase(Container&& shape) : value_(std::move(shape)) { init(); }
+    template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+    explicit ShapeBase(const tt::stl::SmallVector<T>& shape) : value_(shape.begin(), shape.end()) {
+        init();
+    }
+    template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+    explicit ShapeBase(const tt::stl::SmallVector<T>&& shape) : value_(shape.begin(), shape.end()) {
+        init();
+    }
```
Comment (omilyutin-tt) [inline tt_metal/api/tt-metalium/shape_base.hpp]: What is the requirement around this one? E.g. do we want to allow creating shapes from small vectors of `size_t`?  `std::is_integral_v` would work for other types that I don't think we want to implicitly convert to `uint32_t` here.

Also, I think it is better to express this via templated iterators, so that we can use `std::vector` or `tt::stl::Span` here. 
```cpp
template<typename It> ShapeBase(It first, It last) : value_(first, last) { init(); }
```

I was going to suggest concepts, but unfortunately there is the requirement of keeping Metal at C++17 :(

File: tt_metal/api/tt-metalium/math.hpp [hunk 11]
```diff
@@ -20,9 +21,11 @@ namespace tt {
  *
  * @note If b is zero, this results in undefined behavior.
  */
-constexpr uint32_t div_up(uint32_t a, uint32_t b) {
+template <typename A, typename B>
```
Comment (omilyutin-tt) [inline tt_metal/api/tt-metalium/math.hpp:24]: These changes are uncomfortable... Do we need these?

File: tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp [hunk 12]
```diff
@@ -589,3 +585,92 @@ INSTANTIATE_TEST_SUITE_P(
             std::make_pair(PhysicalSize{1, 16}, PhysicalSize{1, 16})),  // tile_shape.
         ::testing::Bool()                                               // transpose
         ));
+
+// Note: tuple is used for ::testing::Combine
+using TilizeUntilizeBigBuffersParams = std::tuple<PhysicalSize, TensorLayoutType, TensorLayoutType, bool, bool>;
```
Comment (omilyutin-tt) [inline tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp]: Can you add a comment what bools in the end mean? (`/*transpose_within_face */bool`)

File: tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp [hunk 13]
```diff
@@ -589,3 +585,92 @@ INSTANTIATE_TEST_SUITE_P(
...
+        ASSERT_TRUE(std::equal(input.begin(), input.end(), converted_back_span.begin()));
+    };
+
+    run_for_type(bfloat16{});
+}
+
+class TilizeUntilizeBigBufferTestsFixture : public ::testing::TestWithParam<TilizeUntilizeBigBuffersParams> {};
+class TilizeUntilizeBigBufferShortTestsFixture : public ::testing::TestWithParam<TilizeUntilizeBigBuffersParams> {};
+
+// Disabled by default, since they take a lot of timne to run. To still run them, use --gtest_also_run_disabled_tests
```
Comment (omilyutin-tt) [inline tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp:649]: Is it possible to run a simplified version of the test, perhaps just one combination of inputs, to validate the computation? If this does not run on CI, we might as well not submit this code :(

Comment (philei-tt) [reply tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp:649]: Yes, that's exactly what I did, just a few lines below this one (TilizeUntilizeBigBufferShortTestsFixture)

File: tt_metal/api/tt-metalium/math.hpp (hunk 11 above)
Comment (philei-tt) [reply tt_metal/api/tt-metalium/math.hpp:24]: There are a few places that put mixed types in those functions, and it was implicitly casting to uint32_t, which caused integer overflow. I thought that casting to common_type will help to prevent such things in the future

File: tt_metal/api/tt-metalium/shape_base.hpp (hunk 10 above)
Comment (philei-tt) [reply tt_metal/api/tt-metalium/shape_base.hpp]: Yes, I changed tt::tt_metal::compute_strides to return SmallVector<size_t> to fix overflow, but Tensor::strides wraps strides into ttnn::Shape... 

I like your proposal, will do

Comment (philei-tt) [reply tt_metal/api/tt-metalium/shape_base.hpp]: Updated. But I had to add change `template<typename It> ` to `template <std::input_iterator It>` because MeshShape has constructor with two integers that conflicted with this

File: tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp (hunk 13 above)
Comment (omilyutin-tt) [reply tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp:649]: Ok got it!

**Review by nvelickovicTT — APPROVED** [review]

**Review by omilyutin-tt — APPROVED** [review]

File: tt_metal/tt_metal.cpp [hunk 14]
```diff
@@ -565,17 +565,17 @@ void WriteToBuffer(Buffer& buffer, tt::stl::Span<const uint8_t> host_buffer) {
 }
 
 void ReadFromDeviceInterleavedContiguous(const Buffer& buffer, uint8_t* host_buffer) {
-    uint32_t page_size = buffer.page_size();
-    uint32_t num_pages = buffer.num_pages();
+    size_t page_size = buffer.page_size();
```
Comment (tt-aho) [inline tt_metal/tt_metal.cpp:565]: Should we be using DeviceAddr here if we're expecting to use a strong type in the future to preserve the type?

**Review by tt-aho — APPROVED** [review]

**Review by tt-asaigal — APPROVED** [review]

**Review by amahmudTT — APPROVED** [review]

File: tt_metal/api/tt-metalium/math.hpp
Comment (blozano-tt) [inline tt_metal/api/tt-metalium/math.hpp:1]: I've always disliked that this file is in `metal` public API.

Why should metal have to tell the world how to do math?

Is there no standard library solution to doing these things?

Maybe we could move this code to `ttsl` ?

**Review by blozano-tt — APPROVED** [review]

File: tt_metal/api/tt-metalium/math.hpp
Comment (philei-tt) [reply tt_metal/api/tt-metalium/math.hpp:1]: Makes sense, but outside the scope of this PR though
